velox/exec/fuzzer/WriterFuzzer.cpp

```diff
@@ -0,0 +1,427 @@
+  // X         null          X
+  // null       X            X
+  // null       X            X
+  // Extract inserted rows from velox execution result.
+  std::vector<VectorPtr> insertedRows = {
+      resultOrError.result->childAt(0)->slice(0, 1)};
+  auto result = std::make_shared<RowVector>(
+      pool_.get(),
+      ROW({"count"}, {insertedRows[0]->type()}),
+      nullptr,
+      1,
+      insertedRows);
+
+  // Presto sql only produces one row which is how many rows are inserted.
+  VELOX_CHECK(
+      expectedResult.size() == 1,
+      "Query returned unexpected result in the reference DB");
+  VELOX_CHECK(
+      assertEqualResults(expectedResult, plan->outputType(), {result}),
+      "Velox and reference DB results don't match");
+
+  // 2. Verify directory layout.
+  // Query Presto to find out table's location on disk.
+  auto path = referenceQueryRunner_->execute("SELECT \"$path\" FROM tmp_write");
+  auto tableDirectoryPath =
+      fs::path(extractSingleValue<StringView>(path)).parent_path();
+  for (int i = 0; i < partitionKeys.size(); i++) {
+    tableDirectoryPath = tableDirectoryPath.parent_path();
+  }
+  VELOX_CHECK(
+      comparePartitions(outputDirectoryPath, tableDirectoryPath),
+      "Velox and reference DB output directory hierarchies don't match");
+
+  LOG(INFO) << "Verified results against reference DB";
+}
+
+fuzzer::ResultOrError WriterFuzzer::execute(
+    const core::PlanNodePtr& plan,
+    int32_t maxDrivers) {
+  LOG(INFO) << "Executing query plan: " << std::endl
+            << plan->toString(true, true);
+  fuzzer::ResultOrError resultOrError;
+  try {
+    AssertQueryBuilder builder(plan);
+    resultOrError.result =
+        builder.maxDrivers(maxDrivers).copyResults(pool_.get());
+  } catch (VeloxUserError& e) {
+    // NOTE: velox user exception is accepted as it is caused by the invalid
+    // fuzzer test inputs.
+    resultOrError.exceptionPtr = std::current_exception();
+  } catch (VeloxRuntimeError& e) {
+    throw e;
+  }
+  return resultOrError;
+}
+
+bool WriterFuzzer::comparePartitions(
+    const std::string& outputDirectoryPath,
+    const std::string& referenceOutputDirectoryPath) {
+  auto partitions = getPartitionNames(outputDirectoryPath);
+  auto referencedPartitions = getPartitionNames(referenceOutputDirectoryPath);
+  LOG(INFO) << "Velox written partitions:" << std::endl;
+  for (std::string p : partitions) {
+    LOG(INFO) << p << std::endl;
+  }
+
+  LOG(INFO) << "Presto written partitions:" << std::endl;
+  for (std::string p : referencedPartitions) {
+    LOG(INFO) << p << std::endl;
+  }
+
+  return partitions == referencedPartitions;
+}
+
+std::set<std::string> WriterFuzzer::getPartitionNames(
+    std::string outputDirectoryPath) {
+  auto fileSystem = filesystems::getFileSystem("/", nullptr);
+  auto directories = fileSystem->listFolders(outputDirectoryPath);
+  std::set<std::string> partitionNames;
+
+  for (std::string directory : directories) {
```

Comment:
```
const std::string& directory
```

Reply:
it would be modified so can't be marked as `const`